File: backend/model/db/operations/telegram_operations.py

from typing import List

from model.db.pools import postgres_db_pool


def get_subscribed_users() -> List[int]:
    try:
        with postgres_db_pool().connection() as conn:
            with conn.cursor() as cur:
                cur.execute("SELECT telegram_user_id FROM Analytics.telegram_receiver WHERE authenticated AND subscribed", ())

                return list(cur.fetchall())

    except Exception as e:
        print("[ERROR][TELEGRAM][DB]Failed to get auth'd users with error=",str(e))
        return []


def auth_chat(chat_id: int, auth: bool, subscribed: bool) -> bool:
    try:
        with postgres_db_pool().connection() as conn:
            with conn.cursor() as cur:
                cur.execute("""
                    INSERT INTO Analytics.telegram_receiver(telegram_user_id, authenticated, subscribed) 
                    VALUES (%s, %s, %s)
                    ON CONFLICT(telegram_user_id) 
                        DO UPDATE SET authenticated = %s, subscribed = %s WHERE Analytics.telegram_receiver.telegram_user_id = %s;
                """, (chat_id, auth, subscribed, auth, subscribed, chat_id))

        return True

    except Exception as e:
        print("[ERROR][TELEGRAM][DB]Failed to insert auth'd user with error=", str(e))
        return False


def is_auth(chat_id: int) -> bool | None:
    value = None
    try:
        with postgres_db_pool().connection() as conn:
            with conn.cursor() as cur:
                cur.execute ("SELECT authenticated FROM Analytics.telegram_receiver WHERE telegram_user_id = %s", (chat_id,))

                value = cur.fetchone()

                if value is None:
                    return False
                return value[0]

    except Exception as e:
        print("[ERROR][TELEGRAM][DB]Failed to check auth status with error=", str(e))
        value = None

    return value


def is_subscribed(chat_id: int) -> bool | None:
    value = None
    try:
        with postgres_db_pool().connection() as conn:
            with conn.cursor() as cur:
                cur.execute ("SELECT subscribed FROM Analytics.telegram_receiver WHERE telegram_user_id = %s", (chat_id,))

                value = cur.fetchone()

                if value is None:
                    return False
                return value[0]

    except Exception as e:
        print("[ERROR][TELEGRAM][DB]Failed to check subscribe status with error=", str(e))
        value = None

    return value

File: backend_aegis/src/model/db/operations/telegram_operations.rs
use sqlx::FromRow;
use tgbot::types::ChatPeerId;


#[derive(FromRow)]
struct TelegramUser {
    telegram_user_id: i64
}

pub async fn get_subscribed_users(pool: &sqlx::PgPool) -> Vec<i64> {
    let items = sqlx::query_as!(TelegramUser, "SELECT telegram_user_id FROM ClientIdentity.telegram_receiver WHERE authenticated AND subscribed;").fetch_all(pool).await;

    let items = match items {
        Ok(i) => i,
        Err(e) => {
            log::error!("[ERROR][ALERTS][TELEGRAM] Failed to query subscribed users from database with error = e '{e}'. Alert events that depend on this will not be sent");
            return Vec::new()
        }
    };

    items.iter().map(|i| i.telegram_user_id).collect()
}

pub async fn auth_chat(chat: ChatPeerId, auth: bool, subscribed: bool, ack_token: String, pool: &sqlx::PgPool) -> Result<(), ()> {
    let id : i64 = serde_json::json!(chat).as_i64().ok_or(())?;
    let result = sqlx::query!("
        INSERT INTO ClientIdentity.telegram_receiver(telegram_user_id, authenticated, subscribed, ack_token) 
        VALUES ($1, $2, $3, $4)
        ON CONFLICT(telegram_user_id) 
            DO UPDATE SET authenticated = $5, subscribed = $6, ack_token = $7 WHERE ClientIdentity.telegram_receiver.telegram_user_id = $8;
    ", id, auth, subscribed, ack_token, auth, subscribed, ack_token, id).execute(pool).await;

    match result {
        Ok(_) => Ok(()),
        Err(e) => {
            log::error!("[ERROR][ALERTS][TELEGRAM][DB] Failed to update user request to create/update auth status into database, with error = '{e}'");
            Err(())
        }
    }
}

pub async fn subscribe_chat(chat: ChatPeerId, subscribed: bool, pool: &sqlx::PgPool) -> Result<(), ()> {
    let id : i64 = serde_json::json!(chat).as_i64().ok_or(())?;
    let result = sqlx::query!("
        UPDATE ClientIdentity.telegram_receiver SET subscribed = $1 WHERE ClientIdentity.telegram_receiver.telegram_user_id = $2 AND authenticated = TRUE;
    ", subscribed, id).execute(pool).await;

    match result {
        Ok(_) => Ok(()),
        Err(e) => {
            log::error!("[ERROR][ALERTS][TELEGRAM][DB] Failed to update user request to create/update auth status into database, with error = '{e}'");
            Err(())
        }
    }
}

pub async fn ack_token_exists(ack_token: String, pool: &sqlx::PgPool) -> Result<(), ()> {

    let result = sqlx::query!("
        SELECT (1) as found FROM ClientIdentity.ack_tokens WHERE ack_token = $1"
    , ack_token).fetch_one(pool).await;

    match result {
        Ok(_) => Ok(()),
        Err(e) => {
            log::error!("[ERROR][ALERTS][TELEGRAM][DB] Failed to update user request to create/update auth status into database, with error = '{e}'");
            Err(())
        }
    }
}

#[derive(FromRow)]
struct ChatAuth {
    authenticated: Option<bool>,
}
pub async fn is_auth(chat: ChatPeerId, pool: &sqlx::PgPool) -> Result<bool, ()> {
    let id : i64 = serde_json::json!(chat).as_i64().ok_or(())?;
    let result = sqlx::query_as!( ChatAuth,"
        SELECT EXISTS(
            SELECT 1 FROM ClientIdentity.telegram_receiver WHERE telegram_user_id = $1 AND authenticated is TRUE LIMIT 1
        ) as authenticated
        ", id
    ).fetch_one(pool).await;

    match result {
        Ok(result) => Ok(result.authenticated.unwrap_or(false)),
        Err(e) => {
            log::error!("[ERROR][ALERTS][TELEGRAM][DB] Failed to request auth state for user from database, with error = '{e}'");
            Err(())
        }
    }
}

pub async fn is_subscribed(chat: ChatPeerId, pool: &sqlx::PgPool) -> Result<bool, ()> {
    let id : i64 = serde_json::json!(chat).as_i64().ok_or(())?;
    let result = sqlx::query_as!( ChatAuth,"
        SELECT EXISTS(
            SELECT 1 FROM ClientIdentity.telegram_receiver WHERE telegram_user_id = $1 AND subscribed is TRUE LIMIT 1
        ) as authenticated
        ", id
    ).fetch_one(pool).await;

    match result {
        Ok(result) => Ok(result.authenticated.unwrap_or(false)),
        Err(e) => {
            log::error!("[ERROR][ALERTS][TELEGRAM][DB] Failed to request subscription state for user from database, with error = '{e}'");
            Err(())
        }
    }
}